Plugins/GitUIPluginInterfaces/RepositoryHosts/GitProtocol.cs

```diff
@@ -0,0 +1,38 @@
+﻿using System;
+using System.Text.RegularExpressions;
+
+namespace GitUIPluginInterfaces.RepositoryHosts
+{
+    public static class GitProtocol
+    {
+        public const string Ssh = "SSH";
+        public const string Https = "HTTPS";
+        private static readonly Regex SshRegex = new Regex("^([^:@]+@)?([^:@]+:[^/:@][^:@]*)$");
+        private static readonly Regex GitUrlRegex = new Regex(@"^((git|ssh|http(s)?)|(git@[\w\.]+))(:(//)?)([\w\.@\:/\-~]+)(\.git)(/)?$"); // From https://stackoverflow.com/a/22312124/717372
+
+        public static bool IsUrlUsingProtocol(this string url, string protocol)
+        {
+            if (string.IsNullOrWhiteSpace(protocol) || string.IsNullOrWhiteSpace(url))
+            {
+                return false;
+            }
+
+            return url.StartsWith($"{protocol}://", StringComparison.CurrentCultureIgnoreCase);
+        }
+
+        public static bool IsUrlUsingHttps(this string url)
+        {
+            return url.IsUrlUsingProtocol(Https);
+        }
+
+        public static bool IsUrlUsingSsh(this string url)
+        {
+            return url.IsUrlUsingProtocol(Ssh) || SshRegex.IsMatch(url);
+        }
+
+        public static bool IsGitUrl(this string url)
+        {
+            return GitUrlRegex.IsMatch(url);
+        }
+    }
+}
```

Comment:
I have a bit of discomfort with the proposed API.

1. 🤔  `IsUrlUsingProtocol` appears to be only used in this class, and feels like an implementation detail that shouldn't be part of the public API.

4. ⚠️  All input parameters for public API must be validated. The current implementation is prone to [ANE](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex.ismatch?view=netframework-4.6.1#System_Text_RegularExpressions_Regex_IsMatch_System_String_).

6. 💭  Given this class contains a set of extensions, perhaps the following name is more fitting: `public static class GitProtocolExtensions`. 
This however makes protocol definitions look awkward, e.g. `GitProtocolExtensions.Ssh` and `GitProtocolExtensions.Https`. 
Do you think we could define an enum instead? This way we could significantly reduce our public API surface:
    ```
    public enum GitProtocol
    {
        None = 0,
        Http,
        Ssh
    }

    public static class GitProtocolExtensions
    {
        public static bool IsUrlUsingProtocol(this string url, GitProtocol protocol)
        {
            if (string.IsNullOrWhiteSpace(url))
            {
                throw new ArgumentException(nameof(url));
            }

            switch (protocol)
            {
                case GitProtocol.Https:
                {
                    return url.IsUrlUsingProtocol(Https);
                }

                case GitProtocol.Ssh:
                {
                    return url.IsUrlUsingProtocol(Ssh) || SshRegex.IsMatch(url);
                }

                case GitProtocol.Git:
                {
                    return GitUrlRegex.IsMatch(url);
                }

                default:
                {
                    throw new NotSupportedException();
                }
            }
        }
    }
    ```

7. 💭  Is there any reason why we couldn't achieve the desired with [`Uri` class](https://docs.microsoft.com/en-us/dotnet/api/system.uri.-ctor?view=netframework-4.6.1#System_Uri__ctor_System_String_System_UriKind_)?
    ```
    public static class GitProtocolExtensions
    {
        public static bool IsUrlUsingProtocol(this string url, GitProtocol protocol)
        {
            if (string.IsNullOrWhiteSpace(url))
            {
                throw new ArgumentException(nameof(url));
            }

            try 
            {
                var uri = new Uri(url, UriKind.Absolute);
                return string.Equals(uri.Scheme, protocol, StringComparer.InvariantCultureIgnoreCase)
            }
            catch (UriFormatException)
            {
                return false;
            }
        }
    }
    ```

Reply:
I've been thinking overnight, are these really "git protocols" or something else?
```public enum GitProtocol
{
    None = 0,
    Http,
    Ssh
}
```

Reply:
It is correct to say that these are Git protocols. 
There is a special git protocol too as well as local
https://git-scm.com/book/en/v2/Git-on-the-Server-The-Protocols